unit/java_bytecode/java_bytecode_specialize_generics/java_bytecode_specialize_generics.cpp

```diff
@@ -0,0 +1,299 @@
+/*******************************************************************\
+
+ Module: Unit tests for specializing generic methods
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+#include <java_bytecode/java_bytecode_language.h>
+#include <java_bytecode/generate_java_generic_method.h>
+#include <util/namespace.h>
+#include <testing-utils/load_java_class.h>
+#include <testing-utils/require_type.h>
+
+SCENARIO(
+  "java_bytecode_specialize_generics",
+  "[core][java_bytecode][java_bytecode_specialize_generics]")
+{
+  symbol_tablet new_symbol_table=load_java_class("generics", ""
+    "./java_bytecode/java_bytecode_specialize_generics");
+
+  generate_java_generic_methodt instantiate_generic_method{};
+
+
+  GIVEN("Some class files with Generics")
+  {
+    WHEN("Methods with generic signatures")
+    {
+      {
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$bound_element.f:()Ljava/lang/Number;"));
+
+        const symbolt &generic_method_symbol=
+          new_symbol_table.lookup_ref(
+            "java::generics$bound_element.f:()Ljava/lang/Number;");
+
+        const code_typet &code=
+          require_type::require_code_type(generic_method_symbol.type, 1);
+
+        require_type::require_java_generic_parameter_variables(
+          code.return_type(),
+          {"java::generics$bound_element::NUM"});
+
+        THEN("Specialize a method with a generic return type")
+        {
+          const symbolt &specialized_method_symbol=instantiate_generic_method(
+            generic_method_symbol,
+            {
+              {"java::generics$bound_element::NUM",
+                symbol_typet("java::java.lang.Integer")
+              }
+            },
+            new_symbol_table);
+
+          REQUIRE(
+            new_symbol_table.has_symbol(
+              "java::generics$bound_element.f:()Ljava/lang/Number;"
+                "<java::java.lang.Integer>"));
+
+          const code_typet &new_code=require_type::require_code_type(
+            specialized_method_symbol.type, 1);
+
+          require_type::require_java_non_generic_type(
+            new_code.return_type(), {"java::java.lang.Integer"});
+        }
+      }
+
```

Comment:
I think it would be good to use a `WHEN("Parsing a method that has a generic parameter")`